### PR TITLE
KAFKA-5297: Reduce time required for shutdown

### PR DESCRIPTION
**Pull request by ambroff**
Avoid reopening and recreating page mappings for index files during
truncation unless it's required by the platform.

In my tests this increases throughput of Log#close by about 70%. See
KAFKA-5297 for details.

### CODE REVIEW

**Comment by ambroff** [comment]
There are obviously a few test cases that are broken. I'm working on those, I just wanted to post this early to get feedback while I work through the issues with this approach.

File: core/src/main/scala/kafka/log/AbstractIndex.scala [hunk 1]
```diff
@@ -103,21 +103,23 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
    */
   def resize(newSize: Int) {
     inLock(lock) {
-      val raf = new RandomAccessFile(file, "rw")
+      val pos = mmap.position()
```
Comment (becketqin) [inline core/src/main/scala/kafka/log/AbstractIndex.scala]: Looked a little bit into the test failure. The failure is caused by this. When a corrupted index file is found, that index file will be deleted from disk. And the recovery logic will call `index.resize()` which will create a new RandomAccessFile() in this case. But the patch removed this logic, so the deleted index file could not be found during the later on deletion.

The simplest fix would be have the following code piece before line 106.
```
      if (!file.exists()) {
        raf.close()
        raf = new RandomAccessFile(file, "rw")
      }
```

However, I do think the code logic should be more clear and the file creation should be more explicit. But we can do that in a separate patch given it is an existing issue, and this patch provides important performance improvement.

Comment (ambroff) [reply core/src/main/scala/kafka/log/AbstractIndex.scala]: Oh! Thanks Becket, I needed a second set of eyes on that.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5476/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5461/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by becketqin** [comment]
We are going to test the patch internally to verify that it solves the problem before checking it in.

**Comment by ijuma** [comment]
@junrao, this approach means an additional file handle per segment. Do you think this is the way to go or the alternative option described in the JIRA?

**Comment by onurkaraman** [comment]
This PR has a file descriptor leak. Currently the RandomAccessFile is only closed on index close, which only happens when closing a log segment and not while deleting a segment.

The PR should also close the RandomAccessFile while deleting a segment.

**Comment by dhruvilshah3** [comment]
@ambroff would you be willing to revive this PR? I think one of the concerns with the PR in its current form is that it consumes an extra file descriptor per segment, which is not ideal.

One of the inefficiencies I noticed (and that you fixed in the PR) was to avoid having to re-mmap the file. Do you think doing this by itself would be a good win, as opposed to the full set of changes in this PR? I am referring to the following change:

```
-        mmap = raf.getChannel().map(FileChannel.MapMode.READ_WRITE, 0, roundedNewSize)

+      if (OperatingSystem.IS_WINDOWS || mmap.capacity() < roundedNewSize) {
+        mmap = raf.getChannel.map(FileChannel.MapMode.READ_WRITE, 0, roundedNewSize)
       }
```

Some initial testing I did suggests we get about 32% improvement by making this change. Thoughts?

**Pull request closed without merging**